numpy/core/src/umath/ufunc_object.c

```diff
@@ -551,6 +551,207 @@ ufunc_get_name_cstr(PyUFuncObject *ufunc) {
+        else {
+#if PY_VERSION_HEX >= 0x03000000
+            PyErr_Format(PyExc_TypeError,
+                         "'%S' is an invalid keyword to ufunc '%s'",
+                         key, ufunc_get_name_cstr(ufunc));
+#else
+            char *str = PyString_AsString(key);
+            if (str == NULL) {
+                PyErr_Clear();
+                PyErr_SetString(PyExc_TypeError, "invalid keyword argument");
+            }
+            else {
+                PyErr_Format(PyExc_TypeError,
+                             "'%s' is an invalid keyword to ufunc '%s'",
+                             str, ufunc_get_name_cstr(ufunc));
+            }
+#endif
+            goto fail;
+        }
+    }
+    return 0;
+
+fail:
+    /*
+     * If we already treated some keywords, repeat the scan of the dict
+     * for everything done previously and call their converters with value=NULL
+     * to indicate they should XDECREF outputs as needed (similar to what is
+     * done in PyArg_ParseTupleAndKeywords).
+     */
+    if (pos != 0) {
+        Py_ssize_t repeat = 0;
+        while (PyDict_Next(kwds, &repeat, &key, &value) && repeat != pos) {
+            int i;
+            converter *convert;
+            void *output = NULL;
+            npy_intp index = locate_key(kwnames, key);
+            va_start(va, kwnames);
+            for (i = 0; i <= index; i++) {
+                convert = va_arg(va, converter *);
+                output = va_arg(va, void *);
+            }
+            va_end(va);
+            if (output) {
+                convert(NULL, output);
+            }
+        }
+    }
+    return -1;
+}
+
+/*
+ * Converters for use in parsing of keywords arguments.
+ */
+NPY_NO_EXPORT int
+_convert_subok(PyObject *obj, int *subok)
+{
+    if (PyBool_Check(obj)) {
+        *subok = (obj == Py_True);
+        return NPY_SUCCEED;
+    }
+    else if (obj == NULL) {
+        return NPY_SUCCEED;
+    }
+    PyErr_SetString(PyExc_TypeError,
+                    "'subok' must be a boolean");
+    return NPY_FAIL;
+}
+
+NPY_NO_EXPORT int
+_convert_keepdims(PyObject *obj, int *keepdims)
+{
+    if (PyBool_Check(obj)) {
+        *keepdims = (obj == Py_True);
+        return NPY_SUCCEED;
+    }
+    else if (obj == NULL) {
+        return NPY_SUCCEED;
+    }
+    PyErr_SetString(PyExc_TypeError,
+                    "'keepdims' must be a boolean");
+    return NPY_FAIL;
+}
+
+NPY_NO_EXPORT int
+_convert_wheremask(PyObject *obj, PyArrayObject **wheremask)
+{
+    /*
+     * Optimization: where=True is the same as no where argument.
+     * This lets us document True as the default.
+     */
+    if (obj != Py_True) {
+        PyArray_Descr *dtype = PyArray_DescrFromType(NPY_BOOL);
+        if (dtype == NULL) {
+            return NPY_FAIL;
+        }
+        *wheremask = (PyArrayObject *)PyArray_FromAny(obj, dtype, 0, 0, 0, NULL);
+        if ((*wheremask) == NULL) {
+            return NPY_FAIL;
+        }
+    }
+    else if (obj == NULL) {
```

Comment:
This condition is never hit.

Reply:
It is if the parser is calling the function again in the failure path to undo an earlier initialization, i.e., in this case when, say, the first keyword was `where` and the second keyword something that is not allowed.

This behaviour is the same as for `PyArg_ParseArgsAndKeywords`

Reply:
p.s. Added a note about the converters to make this clear.

Reply:
But looking more closely at the array ones, I'm not so sure they are actually written to follow the logic... And indeed, reading the documentation [1] more carefully, a function has to return `Py_CLEANUP_SUPPORTED` and this only works in python 3. I'll rewrite accordingly.

[1] https://docs.python.org/3/c-api/arg.html

Reply:
What I meant here is that you had your ifs the wrong way around.

```C
if (obj != PyTrue) {
}
else if (obj == NULL) {
    // but this can never happen, because PyTrue != NULL
}
```

Reply:
Ah, yes, that was silly!